### ISSUE TITLE
[BUG] Possible race when using SRT_EPOLL_ET with file mode?

### ISSUE THREAD

**Issue by erankor** [Haivision/srt#2179]
**Describe the bug**
I think there may be a race when using non-blocking + edge-triggered (SRT_EPOLL_ET) API in file/stream mode. 

In my test, I'm accepting a connection, and immediately start receiving data. 
The sequence of events is -
1. The connection is accepted using srt_accept, and added to epoll with the flags - SRT_EPOLL_IN|SRT_EPOLL_OUT|SRT_EPOLL_ERR|SRT_EPOLL_ET
2. An SRT_EPOLL_OUT event is fired on the socket, my code reacts to it by reading SRTO_PEERVERSION & SRTO_PAYLOADSIZE, but does not send anything (in the test, this side is only receiving...)
3. An SRT_EPOLL_IN event is fired on the socket, in response to that, the code calls srt_recv. But - srt_recv does not return anything - it returns -1 with SRT_EASYNCRCV. I believe this is not supposed to happen - after getting SRT_EPOLL_IN, I expect to get at least some data.

After that, everything is stuck - no additional SRT_EPOLL_IN events are received, and libsrt starts throwing away data, since my application doesn't call srt_recv any more.
It seems the behavior is not entirely consistent, when debugging with gdb and placing breakpoints, in some cases, it works fine (this is the reason I believe it's a race)
When using live/message mode, it seems to be working ok.

**To Reproduce**
I'm hoping the description will give you a clue to what the issue is... even if I shared my code, I'm not sure it would reproduce for you, as it seems related to timing.

**Expected behavior**
I expected -
1. After SRT_EPOLL_IN, srt_recv will return > 0, at least once
2. Even if I didn't get anything after SRT_EPOLL_IN, I'd expect to get another SRT_EPOLL_IN event, since data is definitely coming in... (log starts filling up with "08:40:50.780898/SRT:RcvQ:w1*E:SRT.xt: IPE: Acknowledged seqno %700790837 outruns the RCV buffer state %700790821 - %700790821")

**Desktop (please provide the following information):**
 - OS: Ubuntu 20
 - SRT Version / commit ID: d9b7988347fdbd4c9e1ae7fdc3b0bb3080aa8ae1

Thank you!

Eran

**Comment by maxsharabayko** [comment]
Hi @erankor 
Could you please recheck with the latest master (after PR #2182)?

**Comment by erankor** [comment]
Thanks @maxsharabayko! 
I tested with latest master, and things look better, but still not 100%.
In my test, I'm sending a file from one server to another, I tested 3 times with the previous revision I was at, and 3 times with latest master. 
With the previous code, it failed on all 3 attempts - didn't get any data at all.
With the latest master, I got "something" on all 3 attempts, but in 2/3 cases, it got stuck after a short while - I got ~70k of data, and then it stopped (I'm using buffers of 64k in my application, probably related...).
When it gets stuck, the behavior is as I previously wrote - I'm getting an SRT_EPOLL_IN event, I call srt_recv, it fails with SRT_EASYNCRCV, and no additional events are received.

**Comment by maxsharabayko** [comment]
Thank you for checking, @erankor!
> I'm using buffers of 64k in my application, probably related...

Do you mean you pass a 64 kbytes buffer to `srt_recv(..)`?
Could you please roughly tell the total size of the file you are transferring, and the sending rate you reach to get an impression on timing?

**Comment by erankor** [comment]
Yes, 64 kbytes on both sides - srt_recv & srt_send.
I'm sending a file 1GB in size, when I tested with the sendfile/recvfile examples, it was delivered in ~40 sec.

Thanks!

**Comment by maxsharabayko** [comment]
ℹ️  **Status:** the issue is reproduced.

The new receiver buffer does not require packets to be acknowledged before they can be read, hence the epoll read-readiness notification in `CUDT::sendCtrlAck()` must be done only if reading is actually available.

I.e. add `isRcvBufferReady()` call (see line [8066](https://github.com/Haivision/srt/blob/master/srtcore/core.cpp#L8066)):

```c++
{
    UniqueLock rdlock (m_RecvLock);
    CSync      rdcond (m_RecvDataCond, rdlock);

#if ENABLE_NEW_RCVBUFFER
    // Locks m_RcvBufferLock, which is unlocked above by InvertedLock un_bufflock.
    // Must check read-readiness under m_RecvLock to protect epoll event from changes in readBuffer()/readMessage()
    if (isRcvBufferReady())
#endif
    {
        if (m_config.bSynRecving)
        {
            // signal a waiting "recv" call if there is any data available
            rdcond.signal_locked(rdlock);
        }

        uglobal().m_EPoll.update_events(m_SocketID, m_sPollID, SRT_EPOLL_IN, true);
    }
}
```

**Comment by maxsharabayko** [comment]
Hi @erankor 
I believe the issue is solved by PR #2183. Please don't hesitate to reopen if needed.
Thanks for your prompt and thorough reporting of the issue!

**Comment by erankor** [comment]
Thanks @maxsharabayko !
Tested now with latest master, ran at least 5 times, works great.
